java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java

```diff
@@ -0,0 +1,194 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight.auth2;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.apache.arrow.flight.CallHeaders;
+import org.apache.arrow.flight.CallStatus;
+import org.apache.arrow.flight.auth2.CallHeaderAuthenticator.AuthResult;
+
+/**
+ * An AuthValidator for Token generation and username/password validation.
+ */
+public class BasicAuthValidator implements BasicCallHeaderAuthenticator.AuthValidator {
+
+  private final CredentialValidator credentialValidator;
+  private final AuthTokenManager authTokenManager;
+
+  /**
+   * Creates a validator with supplied AuthTokenManager to generate token
+   * and CredentialValidator to validate username/password.
+   * @param authTokenManager AuthTokenManager used to generate tokens.
+   * @param credentialValidator CredentialValidator used to validate Credentials the username/password
+   */
+  public BasicAuthValidator(CredentialValidator credentialValidator, AuthTokenManager authTokenManager) {
+    super();
+    this.credentialValidator = credentialValidator;
+    this.authTokenManager = authTokenManager;
+  }
+
+  /**
+   * Interface that this validator delegates for determining if the credentials are valid.
+   */
+  interface CredentialValidator {
+    /**
+     * Validate the supplied credentials (username/password) and return the peer identity.
+     * @param username The username to validate.
+     * @param password The password to validate.
+     * @return The peer identity if the supplied credentials are valid.
+     * @throws Exception If the supplied credentials are not valid.
+     */
+    String validateCredentials(String username, String password) throws Exception;
+  }
+
+  /**
+   * Interface that this validator delegates for generating and validating a token.
+   */
+  interface AuthTokenManager {
+    /**
+     * Generate a token for the supplied username and password.
+     * @param username The username to generate the token for.
+     * @param password The password to generate the token for.
+     * @return The generated token.
+     */
+    String generateToken(String username, String password);
+
+    /**
+     * Validate the supplied token and return the peer identity.
+     * @param token The token to be validated.
+     * @return The peer identity if the supplied token is valid.
+     * @throws Exception If the supplied token is not valid.
+     */
+    String validateToken(String token) throws Exception;
+  }
+
+  @Override
+  public AuthResult validateIncomingHeaders(CallHeaders incomingHeaders) throws Exception {
+    final AuthResult authResult = parseAndValidateBearerHeaders(incomingHeaders);
```

Comment:
I don't think this is the right way to go. This incurs the overhead of checking for bearer tokens even if my server only supports basic auth w/o bearer tokens.

What you should do instead is have a CompositeAuthValidator that tries to use one credential validator first (in this case a bearer token validator) then tries to use the other one (in this case a basic validator).

You could potential chain more composites to support more authentication strategies.

Since the bearer token / basic auth case is common, you should add a static factory function to ServerAuthValidator to more easily create an appropriately configured CompositeAuthValidator.

Reply:
Updated code to use a CompositeAuthValidator as suggested.